business/health.go

```diff
@@ -22,133 +23,73 @@ type HealthService struct {
+func (in *HealthService) GetAppHealth(namespace, app, rateInterval string, queryTime time.Time) (models.AppHealth, error) {
 	var err error
 	promtimer := internalmetrics.GetGoFunctionMetric("business", "HealthService", "GetAppHealth")
 	defer promtimer.ObserveNow(&err)
 
-	var services []v1.Service
-	var ws models.Workloads
-
 	appLabel := config.Get().IstioLabels.AppLabelName
 
-	wg := sync.WaitGroup{}
-	wg.Add(2)
-	errChan := make(chan error, 2)
-
 	selectorLabels := make(map[string]string)
 	selectorLabels[appLabel] = app
 	labelSelector := labels.FormatLabels(selectorLabels)
 
-	go func() {
-		defer wg.Done()
-		var err error
-		services, err = in.k8s.GetServices(namespace, selectorLabels)
-		if err != nil {
-			log.Errorf("Error fetching Services per namespace %s and app %s: %s", namespace, app, err)
-			errChan <- err
-		}
-	}()
-
-	go func() {
-		defer wg.Done()
-		var err error
-		ws, err = fetchWorkloads(in.k8s, namespace, labelSelector)
-		if err != nil {
-			log.Errorf("Error fetching Workloads per namespace %s and app %s: %s", namespace, app, err)
-			errChan <- err
-		}
-	}()
-
-	wg.Wait()
-	if len(errChan) != 0 {
-		err := <-errChan
-		return nil, err
+	ws, err := fetchWorkloads(in.k8s, namespace, labelSelector)
+	if err != nil {
+		log.Errorf("Error fetching Workloads per namespace %s and app %s: %s", namespace, app, err)
+		return models.AppHealth{}, err
 	}
 
-	h := in.getAppHealth(namespace, app, rateInterval, queryTime, services, ws)
-	return &h, nil
+	return in.getAppHealth(namespace, app, rateInterval, queryTime, ws)
 }
 
-func (in *HealthService) getAppHealth(namespace, app, rateInterval string, queryTime time.Time, services []v1.Service, ws models.Workloads) models.AppHealth {
+func (in *HealthService) getAppHealth(namespace, app, rateInterval string, queryTime time.Time, ws models.Workloads) (models.AppHealth, error) {
 	health := models.EmptyAppHealth()
 
-	var wg sync.WaitGroup
-	wg.Add(len(services))
-
-	for _, service := range services {
-		go func(service v1.Service) {
-			defer wg.Done()
-			var ports []int32
-			for _, port := range service.Spec.Ports {
-				ports = append(ports, port.Port)
-			}
-			envoy, _ := in.prom.GetServiceHealth(namespace, service.Name, ports)
-			health.Envoy = append(health.Envoy, models.EnvoyHealthWrapper{
-				Service:            service.Name,
-				EnvoyServiceHealth: envoy,
-			})
-		}(service)
-	}
+	envoy, errEnvoy := in.prom.GetEnvoyMembershipRatio(namespace, app)
+	health.Envoy = envoy
 
 	// Fetch services requests rates
-	health.Requests = in.getAppRequestsHealth(namespace, app, rateInterval, queryTime)
+	rate, errRate := in.getAppRequestsHealth(namespace, app, rateInterval, queryTime)
+	health.Requests = rate
 
 	// Deployment status
 	health.WorkloadStatuses = castWorkloadStatuses(ws)
 
-	wg.Wait()
-	return health
+	return health, util.CoalesceErrors(errEnvoy, errRate)
```

Comment:
I suspect that for non-english speakers, perhaps "CombineErrors" can be more clear :-).
I didn't use "Coalesce" verb before today.

Reply:
To be honest, I've never seen it used elsewhere than in code. Originally from SQL syntax: https://www.w3schools.com/sql/func_mysql_coalesce.asp
and from time to time I also see it used in some libs.
"Combine" wouldn't be very accurate, it's not really a combination, however I can rename `FirstNonNil` if it's better.

Reply:
See also: https://en.wikipedia.org/wiki/Null_coalescing_operator

Reply:
"FirstNonNil" sounds more descriptive to me.
I suggested "combine" as synonym of "coalesce" given by Collins dic.
But as perhaps not everyone has used this operator "FirstNonNil" is clear what it does.